File: backend/src/modules/drouting/drouting.service.ts
import { Injectable, NotFoundException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository, Like } from 'typeorm';
import { DrGateway } from './entities/dr-gateway.entity';
import { DrRule } from './entities/dr-rule.entity';
import { DrGroup } from './entities/dr-group.entity';
import { DrCarrier } from './entities/dr-carrier.entity';
import { PaginationDto, PaginatedResult } from '../../common/dto/pagination.dto';
import { KamailioRpcService } from '../../common/services/kamailio-rpc.service';

@Injectable()
export class DroutingService {
  constructor(
    @InjectRepository(DrGateway)
    private gatewayRepository: Repository<DrGateway>,
    @InjectRepository(DrRule)
    private ruleRepository: Repository<DrRule>,
    @InjectRepository(DrGroup)
    private groupRepository: Repository<DrGroup>,
    @InjectRepository(DrCarrier)
    private carrierRepository: Repository<DrCarrier>,
    private kamailioRpcService: KamailioRpcService,
  ) {}

  // ========== Gateway ==========
  async findAllGateways(paginationDto: PaginationDto): Promise<PaginatedResult<DrGateway>> {
    const { page, limit, keyword } = paginationDto;
    const skip = (page - 1) * limit;
    const where: any = {};
    if (keyword) {
      where.gwid = Like(`%${keyword}%`);
    }
    const [items, total] = await this.gatewayRepository.findAndCount({
      where, skip, take: limit, order: { id: 'ASC' },
    });
    return new PaginatedResult(items, total, page, limit);
  }

  async createGateway(data: Partial<DrGateway>): Promise<DrGateway> {
    const gateway = this.gatewayRepository.create(data);
    return this.gatewayRepository.save(gateway);
  }

  async updateGateway(id: number, data: Partial<DrGateway>): Promise<DrGateway> {
    const gateway = await this.gatewayRepository.findOne({ where: { id } });
    if (!gateway) throw new NotFoundException('网关不存在');
    Object.assign(gateway, data);
    return this.gatewayRepository.save(gateway);
  }

  async removeGateway(id: number): Promise<void> {
    const gateway = await this.gatewayRepository.findOne({ where: { id } });
    if (!gateway) throw new NotFoundException('网关不存在');
    await this.gatewayRepository.remove(gateway);
  }

  // ========== Rule ==========
  async findAllRules(paginationDto: PaginationDto): Promise<PaginatedResult<DrRule>> {
    const { page, limit, keyword } = paginationDto;
    const skip = (page - 1) * limit;
    const where: any = {};
    if (keyword) {
      where.prefix = Like(`%${keyword}%`);
    }
    const [items, total] = await this.ruleRepository.findAndCount({
      where, skip, take: limit, order: { priority: 'DESC', ruleid: 'ASC' },
    });
    return new PaginatedResult(items, total, page, limit);
  }

  async createRule(data: Partial<DrRule>): Promise<DrRule> {
    const rule = this.ruleRepository.create(data);
    return this.ruleRepository.save(rule);
  }

  async updateRule(id: number, data: Partial<DrRule>): Promise<DrRule> {
    const rule = await this.ruleRepository.findOne({ where: { ruleid: id } });
    if (!rule) throw new NotFoundException('规则不存在');
    Object.assign(rule, data);
    return this.ruleRepository.save(rule);
  }

  async removeRule(id: number): Promise<void> {
    const rule = await this.ruleRepository.findOne({ where: { ruleid: id } });
    if (!rule) throw new NotFoundException('规则不存在');
    await this.ruleRepository.remove(rule);
  }

  // ========== Group ==========
  async findAllGroups(paginationDto: PaginationDto): Promise<PaginatedResult<DrGroup>> {
    const { page, limit, keyword } = paginationDto;
    const skip = (page - 1) * limit;
    const where: any = {};
    if (keyword) {
      where.username = Like(`%${keyword}%`);
    }
    const [items, total] = await this.groupRepository.findAndCount({
      where, skip, take: limit, order: { groupid: 'ASC', id: 'ASC' },
    });
    return new PaginatedResult(items, total, page, limit);
  }

  async createGroup(data: Partial<DrGroup>): Promise<DrGroup> {
    const group = this.groupRepository.create(data);
    return this.groupRepository.save(group);
  }

  async updateGroup(id: number, data: Partial<DrGroup>): Promise<DrGroup> {
    const group = await this.groupRepository.findOne({ where: { id } });
    if (!group) throw new NotFoundException('分组不存在');
    Object.assign(group, data);
    return this.groupRepository.save(group);
  }

  async removeGroup(id: number): Promise<void> {
    const group = await this.groupRepository.findOne({ where: { id } });
    if (!group) throw new NotFoundException('分组不存在');
    await this.groupRepository.remove(group);
  }

  // ========== Carrier ==========
  async findAllCarriers(paginationDto: PaginationDto): Promise<PaginatedResult<DrCarrier>> {
    const { page, limit, keyword } = paginationDto;
    const skip = (page - 1) * limit;
    const where: any = {};
    if (keyword) {
      where.carrierid = Like(`%${keyword}%`);
    }
    const [items, total] = await this.carrierRepository.findAndCount({
      where, skip, take: limit, order: { id: 'ASC' },
    });
    return new PaginatedResult(items, total, page, limit);
  }

  async createCarrier(data: Partial<DrCarrier>): Promise<DrCarrier> {
    const carrier = this.carrierRepository.create(data);
    return this.carrierRepository.save(carrier);
  }

  async updateCarrier(id: number, data: Partial<DrCarrier>): Promise<DrCarrier> {
    const carrier = await this.carrierRepository.findOne({ where: { id } });
    if (!carrier) throw new NotFoundException('运营商不存在');
    Object.assign(carrier, data);
    return this.carrierRepository.save(carrier);
  }

  async removeCarrier(id: number): Promise<void> {
    const carrier = await this.carrierRepository.findOne({ where: { id } });
    if (!carrier) throw new NotFoundException('运营商不存在');
    await this.carrierRepository.remove(carrier);
  }

  // ========== RPC ==========
  async reload(): Promise<void> {
    await this.kamailioRpcService.call('drouting.reload');
  }

  async getStats(): Promise<any> {
    const gateways = await this.gatewayRepository.count();
    const rules = await this.ruleRepository.count();
    const groups = await this.groupRepository.count();
    const carriers = await this.carrierRepository.count();
    return { gateways, rules, groups, carriers };
  }
}

File: kam-admin-module-routing/src/main/java/io/kamailio/admin/modules/drouting/service/DroutingService.java
package io.kamailio.admin.modules.drouting.service;

import io.kamailio.admin.common.RpcTimeouts;
import io.kamailio.admin.common.dto.PaginatedResult;
import io.kamailio.admin.common.dto.PaginationDto;
import io.kamailio.admin.common.service.KamailioRpcService;
import io.kamailio.admin.modules.drouting.entity.DrCarrier;
import io.kamailio.admin.modules.drouting.entity.DrGateway;
import io.kamailio.admin.modules.drouting.entity.DrGroup;
import io.kamailio.admin.modules.drouting.entity.DrRule;
import io.kamailio.admin.modules.drouting.repository.DrCarrierRepository;
import io.kamailio.admin.modules.drouting.repository.DrGatewayRepository;
import io.kamailio.admin.modules.drouting.repository.DrGroupRepository;
import io.kamailio.admin.modules.drouting.repository.DrRuleRepository;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Map;

@Service
public class DroutingService {
    private final DrGatewayRepository gatewayRepository;
    private final DrRuleRepository ruleRepository;
    private final DrGroupRepository groupRepository;
    private final DrCarrierRepository carrierRepository;
    private final KamailioRpcService kamailioRpc;

    public DroutingService(DrGatewayRepository gatewayRepository, DrRuleRepository ruleRepository,
                           DrGroupRepository groupRepository, DrCarrierRepository carrierRepository,
                           KamailioRpcService kamailioRpc) {
        this.gatewayRepository = gatewayRepository;
        this.ruleRepository = ruleRepository;
        this.groupRepository = groupRepository;
        this.carrierRepository = carrierRepository;
        this.kamailioRpc = kamailioRpc;
    }

    public PaginatedResult<DrGateway> findAllGateways(PaginationDto dto) {
        var page = dto.getKeyword() != null && !dto.getKeyword().isBlank()
                ? gatewayRepository.findByGwidContaining(dto.getKeyword(), PageRequest.of(dto.pageOrDefault() - 1, dto.limitOrDefault(), Sort.by("id")))
                : gatewayRepository.findAll(PageRequest.of(dto.pageOrDefault() - 1, dto.limitOrDefault(), Sort.by("id")));
        return new PaginatedResult<>(page.getContent(), page.getTotalElements(), dto.pageOrDefault(), dto.limitOrDefault());
    }
    @Transactional
    public DrGateway createGateway(DrGateway data) {
        return gatewayRepository.save(data);
    }
    public DrGateway updateGateway(Integer id, DrGateway data) {
        var g = gatewayRepository.findById(id).orElseThrow(() -> new RuntimeException("网关不存在"));
        if (data.getGwid() != null) g.setGwid(data.getGwid());
        if (data.getType() != null) g.setType(data.getType());
        if (data.getAddress() != null) g.setAddress(data.getAddress());
        if (data.getStrip() != null) g.setStrip(data.getStrip());
        if (data.getPriPrefix() != null) g.setPriPrefix(data.getPriPrefix());
        if (data.getAttrs() != null) g.setAttrs(data.getAttrs());
        if (data.getProbeMode() != null) g.setProbeMode(data.getProbeMode());
        if (data.getState() != null) g.setState(data.getState());
        if (data.getSocket() != null) g.setSocket(data.getSocket());
        if (data.getDescription() != null) g.setDescription(data.getDescription());
        return gatewayRepository.save(g);
    }
    @Transactional
    public void removeGateway(Integer id) {
        gatewayRepository.delete(gatewayRepository.findById(id).orElseThrow(() -> new RuntimeException("网关不存在")));
    }

    public PaginatedResult<DrRule> findAllRules(PaginationDto dto) {
        var page = dto.getKeyword() != null && !dto.getKeyword().isBlank()
                ? ruleRepository.findByPrefixContaining(dto.getKeyword(), PageRequest.of(dto.pageOrDefault() - 1, dto.limitOrDefault(), Sort.by(Sort.Direction.DESC, "priority").and(Sort.by("ruleid"))))
                : ruleRepository.findAll(PageRequest.of(dto.pageOrDefault() - 1, dto.limitOrDefault(), Sort.by(Sort.Direction.DESC, "priority").and(Sort.by("ruleid"))));
        return new PaginatedResult<>(page.getContent(), page.getTotalElements(), dto.pageOrDefault(), dto.limitOrDefault());
    }
    @Transactional
    public DrRule createRule(DrRule data) { return ruleRepository.save(data); }
    public DrRule updateRule(Integer id, DrRule data) {
        var r = ruleRepository.findById(id).orElseThrow(() -> new RuntimeException("规则不存在"));
        if (data.getGroupid() != null) r.setGroupid(data.getGroupid());
        if (data.getPrefix() != null) r.setPrefix(data.getPrefix());
        if (data.getTimerec() != null) r.setTimerec(data.getTimerec());
        if (data.getPriority() != null) r.setPriority(data.getPriority());
        if (data.getRouteid() != null) r.setRouteid(data.getRouteid());
        if (data.getGwlist() != null) r.setGwlist(data.getGwlist());
        if (data.getSortAlg() != null) r.setSortAlg(data.getSortAlg());
        if (data.getSortProfile() != null) r.setSortProfile(data.getSortProfile());
        if (data.getAttrs() != null) r.setAttrs(data.getAttrs());
        if (data.getDescription() != null) r.setDescription(data.getDescription());
        return ruleRepository.save(r);
    }
    @Transactional
    public void removeRule(Integer id) {
        ruleRepository.delete(ruleRepository.findById(id).orElseThrow(() -> new RuntimeException("规则不存在")));
    }

    public PaginatedResult<DrGroup> findAllGroups(PaginationDto dto) {
        var page = dto.getKeyword() != null && !dto.getKeyword().isBlank()
                ? groupRepository.findByUsernameContaining(dto.getKeyword(), PageRequest.of(dto.pageOrDefault() - 1, dto.limitOrDefault(), Sort.by("groupid", "id")))
                : groupRepository.findAll(PageRequest.of(dto.pageOrDefault() - 1, dto.limitOrDefault(), Sort.by("groupid", "id")));
        return new PaginatedResult<>(page.getContent(), page.getTotalElements(), dto.pageOrDefault(), dto.limitOrDefault());
    }
    @Transactional
    public DrGroup createGroup(DrGroup data) { return groupRepository.save(data); }
    public DrGroup updateGroup(Integer id, DrGroup data) {
        var g = groupRepository.findById(id).orElseThrow(() -> new RuntimeException("分组不存在"));
        if (data.getUsername() != null) g.setUsername(data.getUsername());
        if (data.getDomain() != null) g.setDomain(data.getDomain());
        if (data.getGroupid() != null) g.setGroupid(data.getGroupid());
        if (data.getDescription() != null) g.setDescription(data.getDescription());
        return groupRepository.save(g);
    }
    @Transactional
    public void removeGroup(Integer id) {
        groupRepository.delete(groupRepository.findById(id).orElseThrow(() -> new RuntimeException("分组不存在")));
    }

    public PaginatedResult<DrCarrier> findAllCarriers(PaginationDto dto) {
        var page = dto.getKeyword() != null && !dto.getKeyword().isBlank()
                ? carrierRepository.findByCarrieridContaining(dto.getKeyword(), PageRequest.of(dto.pageOrDefault() - 1, dto.limitOrDefault(), Sort.by("id")))
                : carrierRepository.findAll(PageRequest.of(dto.pageOrDefault() - 1, dto.limitOrDefault(), Sort.by("id")));
        return new PaginatedResult<>(page.getContent(), page.getTotalElements(), dto.pageOrDefault(), dto.limitOrDefault());
    }
    @Transactional
    public DrCarrier createCarrier(DrCarrier data) { return carrierRepository.save(data); }
    public DrCarrier updateCarrier(Integer id, DrCarrier data) {
        var c = carrierRepository.findById(id).orElseThrow(() -> new RuntimeException("运营商不存在"));
        if (data.getCarrierid() != null) c.setCarrierid(data.getCarrierid());
        if (data.getGwlist() != null) c.setGwlist(data.getGwlist());
        if (data.getFlags() != null) c.setFlags(data.getFlags());
        if (data.getSortAlg() != null) c.setSortAlg(data.getSortAlg());
        if (data.getState() != null) c.setState(data.getState());
        if (data.getAttrs() != null) c.setAttrs(data.getAttrs());
        if (data.getDescription() != null) c.setDescription(data.getDescription());
        return carrierRepository.save(c);
    }
    @Transactional
    public void removeCarrier(Integer id) {
        carrierRepository.delete(carrierRepository.findById(id).orElseThrow(() -> new RuntimeException("运营商不存在")));
    }

    public void reload() { kamailioRpc.droutingReload().block(RpcTimeouts.DEFAULT_BLOCK); }
    public Object getGwStatus() { return kamailioRpc.droutingGwStatus().block(RpcTimeouts.DEFAULT_BLOCK); }
    public Object getCarrierStatus() { return kamailioRpc.droutingCarrierStatus().block(RpcTimeouts.DEFAULT_BLOCK); }
    public Map<String, Object> getStats() {
        return Map.of(
                "gateways", gatewayRepository.count(),
                "rules", ruleRepository.count(),
                "groups", groupRepository.count(),
                "carriers", carrierRepository.count()
        );
    }
}
